pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/LookupTransformFunction.java

```diff
@@ -0,0 +1,115 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.operator.transform.function;
+
+import com.google.common.base.Preconditions;
+import java.util.HashMap;
+import org.apache.pinot.core.common.DataSource;
+import org.apache.pinot.core.data.manager.offline.DimensionTableDataManager;
+import org.apache.pinot.core.operator.blocks.ProjectionBlock;
+import org.apache.pinot.core.operator.transform.TransformResultMetadata;
+import org.apache.pinot.core.plan.DocIdSetPlanNode;
+
+import java.util.List;
+import java.util.Map;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.apache.pinot.spi.data.readers.GenericRow;
+import org.apache.pinot.spi.data.readers.PrimaryKey;
+
+
+public class LookupTransformFunction extends BaseTransformFunction {
+    public static final String FUNCTION_NAME = "lookUp";
+
+    // Functions which will evaluate to lookup parameters
+    private TransformFunction _dimTableNameFunction;
+    private TransformFunction _dimColToLookupFunction;
+    private TransformFunction _dimJoinKeyFunction;
+    private TransformFunction _factJoinValueFunction;
+
+    // potential result sets
+    private String[] _stringValuesSV;
+
+    @Override
+    public String getName() {
+        return FUNCTION_NAME;
+    }
+
+    @Override
+    public void init(List<TransformFunction> arguments, Map<String, DataSource> dataSourceMap) {
+
+        // Check that there are correct number of arguments
+        if (arguments.size() < 4) {
+            throw new IllegalArgumentException("At least 4 arguments are required for LOOKUP transform function");
+        }
+
+        _dimTableNameFunction = arguments.get(0);
+        TransformResultMetadata dimTableNameFunctionResultMetadata = _dimTableNameFunction.getResultMetadata();
+        Preconditions.checkState(dimTableNameFunctionResultMetadata.isSingleValue(),
+            "First argument must be a string representing dimension table name");
+        // TODO validate string
+        // TODO validate this dim table exists
+
+        _dimColToLookupFunction = arguments.get(1);
+        TransformResultMetadata dimColToLookupFunctionResultMetadata = _dimColToLookupFunction.getResultMetadata();
+        Preconditions.checkState(dimColToLookupFunctionResultMetadata.isSingleValue(),
+            "Second argument must be a string representing column name do the lookup from");
+         // TODO validate this column exists
+         // TODO actually lookup the data type of the target column
+
+        _dimJoinKeyFunction = arguments.get(2);
+        TransformResultMetadata dimJoinKeyFunctionResultMetadata = _dimJoinKeyFunction.getResultMetadata();
+        Preconditions.checkState(dimJoinKeyFunctionResultMetadata.isSingleValue(),
+            "Third argument must be the join key on dimension table (primary key)");
```

Comment:
how do you support multiple join keys?

Reply:
Details of the UDF implementation are still in progress, but I just checked in multiple join keys part. Please refresh and take a look.